### DISCUSSION TITLE
AC equation in phase field module

### DISCUSSION THREAD

**Discussion by LiuPengPeter** [idaholab/moose#16118] [Q&A Modules: Phase field] [answered]
Hello guys,
I want to use this AC equation, but it is too complex, especially  the x in it.
**AC equation:**
![Selection_004](https://user-images.githubusercontent.com/39820265/98495034-42adb680-2279-11eb-8768-db4477817a49.png)
**Np:**
![Selection_003](https://user-images.githubusercontent.com/39820265/98495122-6c66dd80-2279-11eb-9ffc-18f2f07da4e9.png)
**x:**
![Selection_002](https://user-images.githubusercontent.com/39820265/98495138-78529f80-2279-11eb-9d62-2e6066a9f4c2.png)

The number of terms of this AC equation is related to x.
My scheme is to use for and if in the residual to sum every term of etas( if eta_j>0). 
It's not easy because M and other coefficients are constants but are different according to different phases...So tough. 
What I could think of is to let them be added as arrays in the input file.  While there is only three phases and 3 different values, but for n grains, the array will be n*n.
Like M in .i file:
    M_ij = '7e-8 7e-8 7e-8 7e-8
            7e-8 7e-8 7e-8 e-6
            7e-8 7e-8 7e-8 e-6
            7e-8 e-6 e-6 7e-8'
Could anyone give me some ideas about there coefficients?

**Comment by GiudGiud** [comment, marked as answer]
Answering this for the sake of closing the post since this was years ago.
MOOSE will let you input vectors and arrays in the input parameters. You can then unfold the vector in the matrix of coefficients you need in your user-code. 

With regards to the model being complex, MOOSE can handle arbitrary numbers of variables and fully couple the equation solves. I think you will find this to be an advantage for your problem. 
I would first start with a lot of repetition in the kernels to build the system, then once you have something working and correct if you need to shorten the syntax you can look into either:
- using an action to generate the system
or 
- using array variables and array kernels to have less objects in the input file
